src/uwtools/rocoto.py

```diff
@@ -25,20 +32,113 @@ def _add_jobname(tree: dict) -> None:
             _add_jobname(subtree)
 
 
-# Public functions
-def write_rocoto_xml(input_yaml: str, input_template: str, rendered_output: str) -> None:
+def _add_tasks(
+    input_yaml: OptionalPath = None,
+) -> None:
     """
-    Main entry point.
+    Define "task" elements in the given config tree and request a "jobname" for each.
 
     :param input_yaml: Path to YAML input file.
-    :param input_template: Path to input template file.
-    :param rendered_output: Path to write rendered XML file.
     """
     values = YAMLConfig(input_yaml)
     tasks = values["tasks"]
     if isinstance(tasks, dict):
         _add_jobname(tasks)
 
+
+# Public functions
+def realize_rocoto_xml(
+    input_yaml: OptionalPath = None,
+    rendered_output: OptionalPath = None,
+) -> bool:  # pragma: no cover
+    """
+    Realize the given YAML file to XML, using fixed Rocoto RelaxNG schema templates. Validate both
+    the YAML and the XML. External functions should call this function.
+
+    :param input_yaml: Path to YAML input file.
+    :param input_template: Path to input template file.
+    :param rendered_output: Path to write rendered XML file.
+    :param schema_file: Path to schema file.
+    """
+
+    _add_tasks(input_yaml)
+
+    rocoto_schema = _rocoto_schema()
+    rocoto_template = _rocoto_template()
+
+    # Validate the YAML.
+    if uwtools.config.validator.validate_yaml(config_file=input_yaml, schema_file=rocoto_schema):
+        # Render the template.
+        write_rocoto_xml(
+            input_yaml=input_yaml,
+            input_template=rocoto_template,
+            rendered_output=rendered_output,
+        )
+        # Validate the XML.
+        if validate_rocoto_xml(input_xml=rendered_output, schema_file=rocoto_template):
+            # If no issues were detected, report success.
+            return True
+        logging.error("Rocoto validation errors identified in %s", rendered_output)
+        return False
+    logging.error("YAML validation errors identified in %s", input_yaml)
+    return False
+
+
+def _rocoto_template() -> str:
+    """
+    The path to the file containing the template to validate the config file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return (path / "rocoto.jinja2").as_posix()
+
+
+def _rocoto_schema() -> str:
+    """
+    The path to the file containing the schema to validate the XML file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return (path / "rocoto.jsonschema").as_posix()
+
+
+def validate_rocoto_xml(input_xml: OptionalPath = None, schema_file: OptionalPath = None) -> bool:
+    """
+    Given a rendered XML file, validate it against the Rocoto schema.
+
+    :param input_XML: Path to rendered XML file.
+    :param schema_file: Path to schema file.
+    """
+
+    # Validate the XML.
+    with open(str(schema_file), "r", encoding="utf-8") as f:
```

Comment:
I don't like using `str()` for `with open()` here but I had some issues with `with writable()` that I'm working on

Reply:
Fixed, just had a weird combination of options before that weren't working